bpf/unwinders/shared.h

```diff
@@ -10,128 +10,140 @@
-    if (unwind_state != NULL) {                                                                                                                                \
-      u64 *scount = bpf_map_lookup_or_try_init(&stack_counts, &unwind_state->stack_key, &zero);                                                                \
-      if (scount) {                                                                                                                                            \
-        __sync_fetch_and_add(scount, 1);                                                                                                                       \
-      }                                                                                                                                                        \
-    }                                                                                                                                                          \
-  })
+#define aggregate_stacks()                                                                            \
+    ({                                                                                                \
+        u64 zero = 0;                                                                                 \
+        unwind_state_t *unwind_state = bpf_map_lookup_elem(&heap, &zero);                             \
+        if (unwind_state != NULL) {                                                                   \
+            u64 *scount = bpf_map_lookup_or_try_init(&stack_counts, &unwind_state->stack_key, &zero); \
+            if (scount) {                                                                             \
+                __sync_fetch_and_add(scount, 1);                                                      \
+            }                                                                                         \
+        }                                                                                             \
+    })
+
+// HACK: On failure, bpf_perf_prog_read_value() zeroes the buffer. We ensure that this always
+// fail with a compile time assert that ensures that the struct size is different to the size
+// of the expected structure.
+//
+// By zeroing the stack we will ensure that stack aggregates work more effectively as otherwise
+// previous values past the stack length will hash the stack to a different value in the map.
+#define bpf_large_memzero(_d, _l)                                                                                        \
+    ({                                                                                                                   \
+        _Static_assert(_l != sizeof(struct bpf_perf_event_value), "stack size must be different to the valid argument"); \
+        bpf_perf_prog_read_value(ctx, _d, _l);                                                                           \
+    })
```

Comment:
You could compute the size in this macro (e.g. `u64 _l = sizeof(*_d);`) rather than forcing it to be passed in.

Reply:
I'll leave it like this in the hopes that it gets changed to some compiler/bpf helper in the future.